pageserver/src/tenant/timeline.rs

```diff
@@ -770,17 +946,43 @@ impl Timeline {
                 "initial size calculation",
                 false,
                 async move {
-                    let calculated_size = self_clone.calculate_logical_size(init_lsn)?;
-                    let result = spawn_blocking(move || {
-                        self_clone.current_logical_size.initial_logical_size.set(calculated_size)
-                    }).await?;
-                    match result {
-                        Ok(()) => info!("Successfully calculated initial logical size"),
-                        Err(existing_size) => error!("Tried to update initial timeline size value to {calculated_size}, but the size was already set to {existing_size}, not changing"),
+                    let mut timeline_state_updates = self_clone.subscribe_for_state_updates();
+                    tokio::select! {
```

Comment:
Makes me scared of cancellation. It's probably OK in this case, logical size calculation shouldn't have any side-effects. Nevertheless, it scares me.

Reply:
I treat panics similar way, so not much to add on my plate, but we can see how it goes.
So far, this one seems to be cancellation safe indeed, since it only reads layers, increments digits and swaps the final result atomically.

Reply:
spawn_blocking cannot be truly cancelled, so may not worth handling currently

Reply:
True, but `select!` could unblock on the status change and set the atomic, checked in the calculation task from time to time.
So I think it works, even though a bit twisted.